### DISCUSSION TITLE
how to compute a strain wich is about hydrostaic stress

### DISCUSSION THREAD

**Discussion by wangzhaohao** [idaholab/moose#27834] [Q&A Modules: Solid mechanics] [answered]
Hello, I want to compute a strain wich is about hydrostaic stress.
so I write this class
```C++
  _stress(this->template getMaterialPropertyByName<RankTwoTensor>("stress")),

    // theoretical density fraction perecent
    if (_geom_type == OxidalDensificationTempl<is_ad>::PLATE)
      _hydro_stress = _stress[_qp].trace() / 3.0;
    else if (_geom_type == OxidalDensificationTempl<is_ad>::CYLINDER)
      _hydro_stress = (_stress[_qp](0, 0) + _stress[_qp](1, 1)) / 2.0;
  ```
and when i test this, there is somthing wrong 
```Assembly
densification <- all_strain <- stress <- densification
```
My input .i is like this
```Assembly
[Materials]
  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 2e11
    poissons_ratio = 0.3
  []
  [density]
    type = Density
    density = 10e3
  []
  [densification]
    type = OxidalDensification
    burnup = burnup
    temperature = temp
    densification_model = MATPRO
    plutonium_concentration = plutonium_concentration
    grain_size = grain_size
    OM_ratio = OM_ratio
    density = density
    eigenstrain_name = 'denficaton_eigenstrain'
  [] 
  [stress]
    type = ComputeFiniteStrainElasticStress
  []
[]
```
and I find some useful code in there [ADViscoplasticityStressUpdate](https://github.com/idaholab/moose/blob/0445d490c50428b8ce3cbf4f27e1b017456306ff/modules/solid_mechanics/src/materials/ADViscoplasticityStressUpdate.C#L66)
so I need to pubic parent class like ADViscoplasticityStressUpdateBase?
I also find the hydrostatic stress is been defined in RankTwoScalarAux, so I also write the other code
```C++
_hydro_stress(this->template coupledValue("hydro_stress"))
```
My question is What is the difference between the latter and the former？
Second question is about creep in https://github.com/idaholab/moose/discussions/27783 
why MOOSE [PowerLawCreepUpdateStress](https://mooseframework.inl.gov/source/materials/PowerLawCreepStressUpdate.html) use the 
![image](https://github.com/idaholab/moose/assets/51054734/4bbd8fe1-effb-41c7-b13e-547b3e6cb283)
Can I think of the formula in the box as von Mises? if not, what are the advantages of doing this?

Thanks for your reply

**Comment by GiudGiud** [comment]
Hello

For the first question, do you mean that the code complains about a circular dependency?
```
densification <- all_strain <- stress <- densification
```
if so then one option would be to move your code into the strain material.

For the second question,
@tophmatthews @jasondhales

**Reply by wangzhaohao** [reply]
> if so then one option would be to move your code into the strain material.
> 
thanks. yes there has a circular wrong.
I don't sure how to do **move code into strain material**. Is there has similar code?
I use a auxvariable to get hydrosticity stress with ranktwoscalaraux. later i post my code.

**Reply by wangzhaohao** [reply]
In both code i public `ComputeEigenstrainBaseTempl`
first code has the circular dependency
```C++
  _stress(this->template getMaterialPropertyByName<RankTwoTensor>("stress")),

template <bool is_ad>
void
OxidalDensificationTempl<is_ad>::computeProperties()
{
    if (_geom_type == OxidalDensificationTempl<is_ad>::PLATE)
      _hydro_stress = _stress[_qp].trace() / 3.0;
    else if (_geom_type == OxidalDensificationTempl<is_ad>::CYLINDER)
      _hydro_stress = (_stress[_qp](0, 0) + _stress[_qp](1, 1)) / 2.0;
}
```
whith test.i
```Assembly
[Materials]
  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 2e11
    poissons_ratio = 0.3
  []
  [density]
    type = Density
    density = 10e3
  []
  [densification]
    type = OxidalDensification
    burnup = burnup
    temperature = temp
    densification_model = MATPRO
    plutonium_concentration = plutonium_concentration
    grain_size = grain_size
    OM_ratio = OM_ratio
    density = density
    eigenstrain_name = 'denficaton_eigenstrain'
  [] 
  [stress]
    type = ComputeFiniteStrainElasticStress
  []
[]
```
second way code is
```C++
   _hydro_stress(this->template coupledValue("hydro_stress")),
```
with test.i is
```Assembly
  [hydro_stress]
    type = RankTwoScalarAux
    variable = hydro_stress
    rank_two_tensor = stress
    scalar_type = hydrostatic
  []
  [densification]
    type = OxidalDensification
    burnup = burnup
    temperature = temp
    densification_model = MATPRO
    plutonium_concentration = plutonium_concentration
    grain_size = grain_size
    OM_ratio = OM_ratio
    density = density
    eigenstrain_name = 'denficaton_eigenstrain'
    hydro_stress = hydro_stress
  [] 
  [stress]
    type = ComputeFiniteStrainElasticStress
  []

```

**Reply by GiudGiud** [reply, marked as answer]
by move code I mean move the definition and computation of the material property.

Thinking about it, it might not help. The properties should not define on each other like this. And if they do, you have to find a way to break the dependency.
Maybe consider lagging one term? Using the value from the previous time step

**Reply by wangzhaohao** [reply]
Hello, I do as you advice
in the code 
```C++
_stress_old(this->template getMaterialPropertyOldByName<RankTwoTensor>("stress"))
```
the result show is same with second way above.
![image](https://github.com/idaholab/moose/assets/51054734/703a962f-b0ff-400a-a02a-fbc8e8957d42)

**Reply by GiudGiud** [reply]
> the result show is same with second way above.

So if the problem is transient then using an old material property just turned whichever time integration scheme into a explicit-implicit scheme. It likely is now 1st order in time (if it was not already)
It is not necessarily an issue.
